src/java/htsjdk/samtools/cram/build/CompressionHeaderFactory.java

```diff
@@ -252,91 +237,65 @@ public int compare(final byte[] o1, final byte[] o2) {
                     dic[i][idIndex][1] = idsAsBytes[j++];
                     dic[i][idIndex][2] = idsAsBytes[j++];
                 }
-                calculator.add(i, map.get(idsAsBytes).value);
                 map.get(idsAsBytes).value = i++;
             }
 
-            calculator.calculate();
-            header.encodingMap.put(EncodingKey.TL_TagIdList,
-                    HuffmanIntegerEncoding.toParam(calculator.values(), calculator.bitLens()));
             header.dictionary = dic;
+
+            header.externalIds.add(exCounter);
+            header.externalCompressors.put(exCounter, ExternalCompressor.createGZIP());
+            header.encodingMap.put(EncodingKey.TL_TagIdList, ExternalIntegerEncoding.toParam(exCounter++));
         }
 
         { // tag values
-            @SuppressWarnings("UnnecessaryLocalVariable") final int unsortedTagValueExternalID = exCounter;
-            header.externalIds.add(unsortedTagValueExternalID);
-            header.externalCompressors.put(unsortedTagValueExternalID,
-                    ExternalCompressor.createRANS(RANS.ORDER.ONE));
-
             final Set<Integer> tagIdSet = new HashSet<Integer>();
+
             for (final CramCompressionRecord record : records) {
-                if (record.tags == null)
+                if (record.tags == null) {
```

Comment:
could record.tags be intialized to an empty array so that you never have to perform this check for null?